include/SKSE/Trampoline.h

```diff
@@ -168,10 +168,111 @@ namespace SKSE
+			static_assert(sizeof(TrampolineAssembly) == 0xA);
+#pragma pack(pop)
+
+			TrampolineAssembly* mem = nullptr;
+			if (const auto it = _5branches.find(a_dst); it != _5branches.end()) {
+				mem = reinterpret_cast<TrampolineAssembly*>(it->second);
+			} else {
+				mem = allocate<TrampolineAssembly>();
+				_5branches.emplace(a_dst, reinterpret_cast<std::byte*>(mem));
+			}
+
+			const auto disp = a_dst - (a_src + sizeof(SrcAssembly));
+			if (!in_range(disp)) {	// the trampoline should already be in range, so this should never happen
+				stl::report_and_fail("displacement is out of range"sv);
+			}
+
+			const auto disp_ret = a_src + sizeof(SrcAssembly) - (reinterpret_cast<std::uint64_t>(mem) + offsetof(TrampolineAssembly, disp) + 4);
+			mem->srcOp1 = *reinterpret_cast<std::uint32_t*>(a_src);
+			mem->srcOp2 = *reinterpret_cast<std::uint8_t*>(a_src + 4);
+			mem->jmp = static_cast<std::uint8_t>(0xE9);
+			mem->disp = static_cast<std::uint32_t>(disp_ret);
+
+			SrcAssembly assembly;
+			assembly.opcode = 0xE9;
+			assembly.disp = static_cast<std::int32_t>(disp);
+			REL::safe_write(a_src, assembly);
+		}
+
+		void write_SafeBranch6(std::uintptr_t a_src, std::uintptr_t a_dst)
+		{
+#pragma pack(push, 1)
+			struct SrcAssembly
+			{
+				std::uint8_t opcode;  // 0 - 0xE9
+				std::int32_t disp;	  // 1
+			};
+			static_assert(offsetof(SrcAssembly, opcode) == 0x0);
+			static_assert(offsetof(SrcAssembly, disp) == 0x1);
+			static_assert(sizeof(SrcAssembly) == 0x5);
+
+			struct TrampolineAssembly
+			{
+				std::uint32_t srcOp1;
+				std::uint16_t srcOp2;
+				std::uint8_t  jmp;
+				std::uint32_t disp;
+			};
+			static_assert(offsetof(TrampolineAssembly, srcOp1) == 0x0);
+			static_assert(offsetof(TrampolineAssembly, srcOp2) == 0x4);
+			static_assert(offsetof(TrampolineAssembly, jmp) == 0x6);
+			static_assert(offsetof(TrampolineAssembly, disp) == 0x7);
+			static_assert(sizeof(TrampolineAssembly) == 0xB);
+#pragma pack(pop)
+
+			TrampolineAssembly* mem = nullptr;
+			if (const auto it = _6branches.find(a_dst); it != _6branches.end()) {
+				mem = reinterpret_cast<TrampolineAssembly*>(it->second);
+			} else {
+				mem = allocate<TrampolineAssembly>();
+				_6branches.emplace(a_dst, reinterpret_cast<std::byte*>(mem));
+			}
+
+			const auto disp = a_dst - (a_src + sizeof(SrcAssembly));
+			if (!in_range(disp)) {	// the trampoline should already be in range, so this should never happen
+				stl::report_and_fail("displacement is out of range"sv);
+			}
+
+			const auto disp_ret = a_src + 6 - (reinterpret_cast<std::uint64_t>(mem) + offsetof(TrampolineAssembly, disp) + 4);
+			mem->srcOp1 = *reinterpret_cast<std::uint32_t*>(a_src);
+			mem->srcOp2 = *reinterpret_cast<std::uint16_t*>(a_src + 4);
+			mem->jmp = 0xE9;
+			mem->disp = static_cast<std::uint32_t>(disp_ret);
+
+			SrcAssembly assembly;
+			assembly.opcode = 0xE9;
+			assembly.disp = static_cast<std::int32_t>(disp);
+			REL::safe_write(a_src, assembly);
+		}
+
```

Comment:
Trampoline already contains branching facilities. These new functions are redundant.

Reply:
The functions automate the return process, how does anything in the trampoline class provide this functionality?

Reply:
If you need to call the original function, then you can store the address returned by existing facilities and call it later. The method you propose assumes the instrumented call site uses a `call` op, and does not preserve arguments for the injected destination. I will not accept this.